### PR TITLE
Extend authentication.py to store tokens across sessions

### PR DESCRIPTION
**Pull request by SirGhostal**
Functions: `authenticate`, `refresh` and `sign_out` now take an optional parameter of a dict.
When a success response is received, the functions will now update the launcher_profiles dict with the newly updated access_token/profile name and uuid.

I also added an Class ProfilesFile() which a user may optionally use as their json dictionary to open and save their launcher_profiles.
This class also generates a version4 uuid for the client_token when no launcher_profiles.json file is found as does the vanilla client.
This client token is specific for that instance of the client and should be used for all future requests with corresponding access tokens. 

Each client only needs one client token and if provided in the payload, the server won't send you a new/different client_token.
See https://wiki.vg/Authentication

### CODE REVIEW

**Comment by Zachy24** [comment]
Requesting a review @joodicator @ammaraskar 
Thanks :-)

File: minecraft/authentication.py [hunk 1]
```diff
@@ -206,6 +267,15 @@ def sign_out(username, password):
                             {"username": username, "password": password})
 
         if _raise_from_response(res) is None:
```
Comment (joodicator) [inline minecraft/authentication.py]: This error was present in the original code (which was apparently never tested), but needs to be fixed anyway: the `signout` and `invalidate` endpoints return a `204 No Content` when successful, but `_raise_from_response` always treats `204` as an error. Therefore, this method always raises an exception after the signout completes normally.

Perhaps `_raise_from_response` should take a keyword argument, e.g. `success_status`, which defaults to `requests.codes.ok` (i.e. `200`), but can be specified as `requests.codes.no_content` (i.e. `204`) in cases like these.

File: minecraft/authentication.py [hunk 2]
```diff
@@ -206,6 +267,15 @@ def sign_out(username, password):
...
+                for entry in database:
+                    if entry['username'] == username:
+                        del database[entry]
+                        break
+
+            self._authenticated = False
+
             return True
 
     def invalidate(self):
```
Comment (joodicator) [inline minecraft/authentication.py]: This method should be updated in the same way that `sign_out` is updated.

**Comment by joodicator** [comment]
Thank you for this pull request. Further to the specific comments I've made on the code, some general comments on design and testing follow:

1. Your implementation requires the user to pass `launcher_profiles_dict` as an argument to each method which might use it, but it seems that you could make things simpler for the user by having the constructor of `AuthenticationToken` take this dict once and store it for subsequent use. I can't imagine that any user would want to use two different dicts with the same token. What do you think?

2. Currently, it seems that the only option the user has for reconstructing an `AuthenticationToken` loaded from a file is to call the constructor with each parameter extracted explicitly from the loaded dict; but it seems like the constructor could easily take care of this for the user by just taking the dict and selecting the correct profile based on the provided username (or also the profile name, if Mojang ever implement multiple profiles).

3. It might be good to amend `start.py` to use this new functionality (with the option to disable it and/or specify the `launcher_profiles.json` path via a command line switch, perhaps), so we have a canonical example of how to use it as intended, and a quick way to manually test that it's working correctly.

4. The JSON format used by the `launcher_profiles_dict` is similar to that used by the vanilla Minecraft launcher's own `launcher_profiles.json`, but deviates some ways, for example using `database` instead of `authenticationDatabase`, and `client_token` instead of `clientToken`. May I ask what the rationale behind these slight changes was? It seems to me that if it used (a subset of) the same format used by the vanilla launcher, it would be more intuitive and less surprising, and (perhaps) allow for interoperation between the two systems.

5. This module is quite an old part of pyCraft which could do with some updating. I hadn't ever looked at it closely before myself, and I found some errors in the existing code as well as in the code you added. There might be more errors that I've missed, so it would be good to have unit tests covering this module more thoroughly, to find such errors. Auto-testing using Mojang's auth servers is not reliable, since they are rate-limited, so something along the lines of [`FakeServer`](https://github.com/ammaraskar/pyCraft/blob/f6f65117881a8132bff1d4846e41420508a23fc4/tests/fake_server.py#L305) might be in order. I'll see about writing something like this when I get the chance, or if someone else wishes to take it on, they are more than welcome to do so. I suggest we wait until we can get better test coverage before merging this PR, but hopefully we can get it merged before 1.13.1 is released.

**Pull request closed without merging**